### PR TITLE
OnBoarding: Bug marker

### PR DESCRIPTION
**Pull request by robertomonteromiguel**
## Motivation
We need a bug marker for onboarding, in order to skip test by virtual machine properties

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

### CODE REVIEW

File: utils/virtual_machine/utils.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        parameters = []
+
+        for vm in getattr(context.scenario, "required_vms", []):
+            bug_found = False
+            if bugs and hasattr(bugs, "__len__") and len(bugs) > 0:
+                for bug in bugs:
+                    if (
+                        ("vm_name" not in bug or vm.name == bug["vm_name"])
+                        and (not "vm_cpu" in bug or vm.os_cpu == bug["vm_cpu"])
+                        and (not "weblog_variant" in bug or context.weblog_variant == bug["weblog_variant"])
```
Comment (szegedi) [inline utils/virtual_machine/utils.py:29]: maybe just call this `weblog`? Not sure if `weblog_variant` conveys additional meaning.

File: utils/virtual_machine/utils.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    def decorator(func):
+
+        parameters = []
+
+        for vm in getattr(context.scenario, "required_vms", []):
+            bug_found = False
+            if bugs and hasattr(bugs, "__len__") and len(bugs) > 0:
+                for bug in bugs:
+                    if (
+                        ("vm_name" not in bug or vm.name == bug["vm_name"])
```
Comment (szegedi) [inline utils/virtual_machine/utils.py:27]:
```suggestion
                        (not "vm_name" in bug or vm.name == bug["vm_name"])
```
for consistency with the other two?

File: utils/virtual_machine/utils.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                        and (not "vm_cpu" in bug or vm.os_cpu == bug["vm_cpu"])
+                        and (not "weblog_variant" in bug or context.weblog_variant == bug["weblog_variant"])
+                    ):
```
Comment (szegedi) [inline utils/virtual_machine/utils.py:30]: Maybe I'm daft, but isn't the predicate almost-but-not-exactly inverted here? I would expect that the condition for marking the test as known failure would be:
```suggestion
                        ("vm_name" in bug and vm.name == bug["vm_name"])
                        or ("vm_cpu" in bug and vm.os_cpu == bug["vm_cpu"])
                        or ("weblog_variant" in bug and context.weblog_variant == bug["weblog_variant"])
                    ):
```
I'm sure it's just me missing something here.

**Review by szegedi — CHANGES_REQUESTED** [review]

File: utils/virtual_machine/utils.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+                        parameters.append(pytest.param(vm, marks=pytest.mark.xfail(reason=f"bug: {bug['reason']}")))
```
Comment (cbeauchesne) [inline utils/virtual_machine/utils.py:31]: Could you add a check on the reason, to ensure it's a JIRA ticket ? 

See https://github.com/DataDog/system-tests/blob/main/utils/_decorators.py#L12

The good way is to  probably to add a function `ensure_jira_ticket(reason:str)` that raise a `ValueError` and reuse it.

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

File: utils/virtual_machine/utils.py (hunk 1 above)
Comment (robertomonteromiguel) [reply utils/virtual_machine/utils.py:29]: I'm using the same name (weblog_variant) as other decorators:
` @flaky(weblog_variant="test-app-java-buildpack", reason="APMON-1595")`

File: utils/virtual_machine/utils.py (hunk 3 above)
Comment (robertomonteromiguel) [reply utils/virtual_machine/utils.py]: I think it's ok. But please, confirm me this after this explanation:

for example:
* mark as bug all machines with cpu==amd64 AND weblog_variant=="xyz"
`bugs=[{"vm_cpu": "amd64", "weblog_variant":"xyz", "reason": "APMON-1576"}]`

* mark as bug all machines with vm_name="vm1" AND cpu==amd64 AND weblog_variant=="xyz"
`bugs=[{"vm_name": "vm1", "vm_cpu": "amd64", "weblog_variant":"xyz", "reason": "APMON-1576"}]`
I mean, if I have a machine "vm1" with cpu "arm64" is going to execute the test

File: utils/virtual_machine/utils.py (hunk 1 above)
Comment (cbeauchesne) [reply utils/virtual_machine/utils.py:30]: I don't know if `weblog` is a better name (probably `variant` is a better one, it's more generic, and this abstraction may not concerns only `weblog`). 

But for this particular PR, I agree with Roberto : continue to use the one used elsewhere to not add any entropy. And if we feel it should change, adress this globally in another PR.

I pick my entire part of responsibility here, I've never settled for a name.

File: utils/virtual_machine/utils.py (hunk 4 above)
Comment (robertomonteromiguel) [reply utils/virtual_machine/utils.py]: done!

File: utils/virtual_machine/utils.py (hunk 1 above)
Comment (szegedi) [reply utils/virtual_machine/utils.py:33]: Makes sense – I didn't realize it is used as it is elsewhere. Thanks for the explanation!

File: utils/virtual_machine/utils.py (hunk 3 above)
Comment (szegedi) [reply utils/virtual_machine/utils.py]: Oh, got it – if there _is_ a bug marker, then the vm needs to match all specified fields to be a known failing one, and it of course implicitly matches any unspecified fields. Thanks for bearing with me!

**Review by szegedi — APPROVED** [review]

**Pull request merged**